File: projects/spring-boot-rest/demo/src/main/java/com/sourabh/demo/controller/CustomerController.java
package com.sourabh.demo.controller;

import com.sourabh.demo.model.Customer;
import com.sourabh.demo.service.CustomerService;
import io.swagger.annotations.ApiOperation;
import org.bson.Document;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

@RestController
@RequestMapping("/api")
public class CustomerController {

    @Autowired
    CustomerService customerService;
    @ApiOperation(value = "Fetches complete customers information.",
            notes = "Gives the complete list of customers who has ever book a room.")
    @GetMapping("hotels/customers/all")
    public List<Customer> getAll(){
        return customerService.getAll();
    }

    @ApiOperation(value = "Shows current occupants count",
            notes = "Shows the summarized details of which all rooms are booked.\n" +
                    "Their current counts.")
    @GetMapping("hotels/customers/occupants")
    public Document getOccupantsCounts(){
        return customerService.getOccupantsCounts();
    }

    @ApiOperation(value = "Shows occupants count",
            notes = "Shows the historical view of which rooms are booked how many times till now.")
    @GetMapping("hotels/customers/occupants/history")
    public Document getOccupantsCountsHistory(){
        return customerService.getOccupantsCountsHistory();
    }

    @ApiOperation(value = "Shows datewise and typewise grouped result",
            notes = "Gives a aggregate of Rooms booked and date on which they were booked.\n" +
                    "Summarizes the result.")
    @GetMapping("hotels/customers/bookings-datewise")
    public  Document getDatewiseBookings(){
        return customerService.getDateWiseBooking();
    }
}

File: projects/spring-boot-rest-kt/hotel-reservation/src/main/kotlin/com/sourabh/hotelreservation/controller/CustomerController.kt
package com.sourabh.hotelreservation.controller



import com.sourabh.hotelreservation.model.Customer
import com.sourabh.hotelreservation.repository.TypeOccupant
import com.sourabh.hotelreservation.service.CustomerService
import io.swagger.annotations.ApiOperation
import org.bson.Document
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.data.mongodb.core.aggregation.AggregationResults
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController


@RestController
@RequestMapping("/api")
class CustomerController(@Autowired val customerService: CustomerService) {
    @GetMapping("/customers/all")
    @ApiOperation(value = "Fetches complete customers information.", notes = "Gives the complete list of customers who has ever book a room.")
    fun getAll():MutableList<Customer>{
        return customerService.getAll()
    }

    @GetMapping("/customers/occupants")
    @ApiOperation(value = "Shows current occupants count", notes = """Shows the summarized details of which all rooms are booked.
Their current counts.""")
    fun getOccupantsCounts(): Document {
        return customerService.getOccupantsCounts()
    }

    @GetMapping("/customers/occupants/history")
    @ApiOperation(value = "Shows occupants count", notes = "Shows the historical view of which rooms are booked how many times till now.")
    fun getOccupantsCountsHistory(): Document {
        return customerService.getOccupantsCountHistory()
    }

    @GetMapping("/customers/bookings-datewise")
    @ApiOperation(value = "Shows datewise and typewise grouped result", notes = "Gives a aggregate of Rooms booked and date on which they were booked.Summarizes the result.")
    fun getDateWiseBookings():Document{
        return customerService.getDateWiseBookings()
    }


}
